File: Python/binarySearchTree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class BinarySeachTree:
    def __init__(self):
        self.root = None

    def insert(self, value):
        newNode = Node(value)
        if self.root == None:
            self.root = newNode
        curr = self.root
        while curr != None:
            if value < curr.value and curr.left != None:
                curr = curr.left
            elif value > curr.value and curr.right != None:
                curr = curr.right
            else:
                break
        if value < curr.value:
            curr.left = newNode
        elif value > curr.value:
            curr.right = newNode
        return "Inserted"
    
    def getSuccessor(self, root):
        curr = root
        while curr != None and curr.left != None:
            curr = curr.left
        return curr
    
    def deleteNode(self, root, value):
        if root == None:
            print("Data not Found")
            return root
        if value < root.value:
            root.left = self.deleteNode(root.left, value)
        elif value > root.value:
            root.right = self.deleteNode(root.right, value)
        else:
            if root.left == None:
                return root.right
            if root.right == None:
                return root.left
            successor = self.getSuccessor(root.right)
            root.value = successor.value
            root.right = self.deleteNode(root.right, root.value)
        return root
    
    def delete(self, value):
        if self.root == None:
            print("Tree is Empty")
        else:
            self.root = self.deleteNode(self.root, value)
            print("Deleted")

    def searchNode(self, root, value):
        if root == None:
            return "Data not Found"
        if value < root.value:
            self.searchNode(root.left, value)
        elif value > root.value:
            self.searchNode(root.right, value)
        else:
            return root.value
    
    def search(self, value):
        if self.root == None:
            return "Tree is Empty"
        else:
            return self.searchNode(self.root, value)
        
    def preorderTraversal(self, root):
        print(root.value)
        if root.left != None:
            self.preorderTraversal(root.left)
        if root.right != None:
            self.preorderTraversal(root.right)
        
    def inorderTraversal(self, root):
        if root.left != None:
            self.inorderTraversal(root.left)
        print(root.value)
        if root.right != None:
            self.inorderTraversal(root.right)
    
    def postorderTraversal(self, root):
        if root.left != None:
            self.postorderTraversal(root.left)
        if root.right != None:
            self.postorderTraversal(root.right)
        print(root.value)

    def traversals(self):
        if self.root == None:
            print("Tree is Empty")
            return
        print("\nPreorder Traversal")
        self.preorderTraversal(self.root)
        print("\nInorder Traversal")
        self.inorderTraversal(self.root)
        print("\nPostorder Traversal")
        self.postorderTraversal(self.root)

binarySearchTree = BinarySeachTree()
while 1:
    print("\n0. Exit\n1. Insert\n2. Delete\n3. Search\n4. Traversals")
    option = int(input("\nOption: "))
    if option == 0:
        break
    match option:
        case 1:
            value = int(input("Value: "))
            print(binarySearchTree.insert(value))
        case 2:
            value = int(input("Value: "))
            binarySearchTree.delete(value)
        case 3:
            value = int(input("Value: "))
            print(binarySearchTree.search(value))
        case 4:
            binarySearchTree.traversals()

File: Java/BinarySearchTree.java
import java.util.Scanner;

class Node {
	public int value;
	public Node left;
	public Node right;

	public Node(int value) {
		this.value = value;
		this.left = null;
		this.right = null;
	}	
}

class BinarySearchTreeDate {
	public Node root;

	public BinarySearchTreeDate() {
		this.root = null;
	}

	public String insert(int value) {
		Node newNode = new Node(value);
		if(this.root == null) {
			this.root = newNode;
			return "Inserted";
		}
		Node curr = this.root;
		while(true) {
			if(value < curr.value && curr.left != null) curr = curr.left;
			else if(value > curr.value && curr.right != null) curr = curr.right;
			else break;
		}
		if(value <= curr.value) {
			curr.left = newNode;
		} else {
			curr.right = newNode;
		}
		return "Inserted";
	}

	public Node getSuccessor(Node curr) {
		while (curr != null && curr.left != null) {
			curr = curr.left;
		}
		return curr;
	}

	public Node deleteNode(Node root, int value) {
		if(root == null) {
			System.out.println("Data not Found");
			return root;
		}
		if(value < root.value) root.left = deleteNode(root.left, value);
		else if(value > root.value) root.right = deleteNode(root.right, value);
		else {
			if (root.left == null) return root.right;
			if (root.right == null) return root.left;
			else {
				Node successor = getSuccessor(root.right);
				root.value = successor.value;
				root.right = deleteNode(root.right, successor.value);
			}
		}
		return root;
	}

	public String delete(int value) {
		if(this.root == null) return "Tree is Empty";
		this.root = deleteNode(this.root, value);
		return "Deleted";
	}

	public int searchNode(Node root, int value) {
		if(root == null) return -1;
		if(value < root.value) return searchNode(root.left, value);
		else if(value > root.value) return searchNode(root.right, value);
		else return root.value;
	}

	public int search(int value) {
		if(this.root == null) {
			System.out.println("Tree is Empty");
			return -1;
		} else {
			return searchNode(this.root, value);
		}
	}

	public void preorderTraversal(Node root) {
		System.out.println(root.value + " ");
		if(root.left != null) preorderTraversal(root.left);
		if(root.right != null) preorderTraversal(root.right);
	}

	public void inorderTraversal(Node root) {
		if(root.left != null) inorderTraversal(root.left);
		System.out.println(root.value + " ");
		if(root.right != null) inorderTraversal(root.right);
	}

	public void postorderTraversal(Node root) {
		if(root.left != null) postorderTraversal(root.left);
		if(root.right != null) postorderTraversal(root.right);
		System.out.println(root.value + " ");
	}

	public void traverse() {
		if(this.root == null) System.out.println("Tree is Empty");
		System.out.println("\nPreorder Traversal");
		preorderTraversal(this.root);
		System.out.println("\nInorder Traversal");
		inorderTraversal(this.root);
		System.out.println("\nPostorder Traversal");
		postorderTraversal(this.root);
	}
}

class BinarySearchTree {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		BinarySearchTreeDate binarySearchTreeDate = new BinarySearchTreeDate();
		int option;
		while (true) {
			System.out.print("\n0. Exit\n1. Insert\n2. Delete\n3. Search\n4. Traverse\nOption: ");
			option = sc.nextInt();
			if (option == 0) return;
			switch (option) {
				case 1 -> {
					System.out.print("value: ");
					int value = sc.nextInt();
					System.out.println(binarySearchTreeDate.insert(value));
				}
				case 2 -> {
					System.out.print("value: ");
					int value = sc.nextInt();
					System.out.println(binarySearchTreeDate.delete(value));
				}
				case 3 -> {
					System.out.print("value: ");
					int value = sc.nextInt();
					System.out.println(binarySearchTreeDate.search(value));
				}
				case 4 -> binarySearchTreeDate.traverse();
				default -> System.out.println("Please Choose a valid option");
			}
		}
	}
}